src/reporting/clients/ReportsClient.ts

```diff
@@ -0,0 +1,141 @@
+        const response: ReportCollection = await this.fetchData<ReportCollection>(url, requestOptions);
+        return {
+          values: response.reports,
+          _links: response._links,
+        }
+    }));
+  }
+
+  public async getReport(accessToken: AccessToken, reportId: string): Promise<Report> {
+    const url = `${this.basePath}/reports/${encodeURIComponent(reportId)}`;
+    const requestOptions: RequestInit = this.createRequest("GET", accessToken);
+    return (await this.fetchData<ReportSingle>(url, requestOptions)).report;
+  }
+
+  public async createReport(accessToken: AccessToken, report: ReportCreate): Promise<Report>{
+    if (!report.displayName) {
+      throw new RequiredError(
+        'displayName',
+        'Required field displayName of report was null or undefined when calling createReport.',
+      );
+    }
+    if (!report.projectId) {
+      throw new RequiredError(
+        'projectId',
+        'Required field of report was null or undefined when calling createReport.',
+      );
+    }
+
+    const url = `${this.basePath}/reports/`;
+    const requestOptions: RequestInit = this.createRequest("POST", accessToken, JSON.stringify(report));
+    return (await this.fetchData<ReportSingle>(url, requestOptions)).report;
+  }
+
+  public async updateReport(accessToken: AccessToken, reportId: string, report: ReportUpdate): Promise<Report> {
+    if (report.deleted == null && report.description == null && report.displayName == null) {
+      throw new RequiredError(
+        'report',
+        'All fields of report were null or undefined when calling updateReport.',
+      );
+    }
+    if (report.displayName === "") {
+      throw new RequiredError(
+        'displayName',
+        'Field display of report was empty when calling createReportMapping.',
+      );
+    }
+
+    const url = `${this.basePath}/reports/${encodeURIComponent(reportId)}`;
+    const requestOptions: RequestInit = this.createRequest("PATCH", accessToken, JSON.stringify(report));
+    return (await this.fetchData<ReportSingle>(url, requestOptions)).report;
+  }
+
+  public async deleteReport(accessToken: AccessToken, reportId: string): Promise<Response> {
+    const url = `${this.basePath}/reports/${encodeURIComponent(reportId)}`;
+    const requestOptions: RequestInit = this.createRequest("DELETE", accessToken);
+    return this.fetchData<Response>(url, requestOptions);
+  }
+
+  public async getReportMappings(accessToken: AccessToken, reportId: string, top?: number): Promise<ReportMapping[]> {
+    const reportMappings: Array<ReportMapping> = [];
+    const reportMappingIterator = this.getReportMappingsIterator(accessToken, reportId, top);
+    for await(const reportMapping of reportMappingIterator) {
+      reportMappings.push(reportMapping);
+    }
+    return reportMappings;
+  }
+
+  public getReportMappingsIterator(accessToken: AccessToken, reportId: string, top?: number): EntityListIterator<ReportMapping> {
```

Comment:
Add validation for `top` if it is given. It could be moved to an utility function, because it will be repeated over and over again. The limits are: `top > 0 && top <= 1000`. When `top` is outside of those limits, the API returns 422 status code (Unprocessable Entity).

Reply:
Added new function to operationsBase called topInRangeValidation. It validates whether the top value is in the required range.

Reply:
Changed it to isValidTop and made custom validations for every function which requires top.